### PR TITLE
Explicit Boundary Histogram Metric Aggregator

### PR DESCRIPTION
**Pull request by victlu**
Simple implementation of an Explicit Boundary Histogram metric aggregator.

### CODE REVIEW

File: src/OpenTelemetry/Metrics/MetricAggregators/HistogramMetricAggregator.cs [hunk 1]
```diff
@@ -62,11 +121,38 @@ internal HistogramMetricAggregator(string name, string description, string unit,
...
+            {
+                val = (long)(object)value;
+            }
+            else if (typeof(T) == typeof(double))
+            {
+                val = (double)(object)value;
+            }
+            else
+            {
+                val = (double)(object)value;
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MetricAggregators/HistogramMetricAggregator.cs:137]: trust the SDK to send only long or double, anything else - ignore/log/throw.

File: src/OpenTelemetry/Metrics/MetricAggregators/HistogramMetricAggregator.cs [hunk 2]
```diff
@@ -22,17 +22,76 @@ namespace OpenTelemetry.Metrics
...
+            this.boundaries = boundaries;
+            this.buckets = this.InitializeBucket(boundaries);
+        }
+
+        internal HistogramBucket[] InitializeBucket(double[] boundaries)
+        {
+            var buckets = new HistogramBucket[boundaries.Length + 1];
+
+            var lastBoundary = boundaries[0];
+            for (int i = 0; i <= boundaries.Length; i++)
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MetricAggregators/HistogramMetricAggregator.cs:59]: <= intentional?

File: test/OpenTelemetry.Tests/Metrics/MetricAPITest.cs [hunk 3]
```diff
@@ -14,20 +14,74 @@
...
+
+            Assert.NotNull(metric);
+            Assert.IsType<HistogramMetricAggregator>(metric);
+
+            if (metric is HistogramMetricAggregator agg)
+            {
+                int len = 0;
+                foreach (var bucket in agg.Buckets)
+                {
+                    Assert.Equal(1, bucket.Count);
```
Comment (cijothomas) [inline test/OpenTelemetry.Tests/Metrics/MetricAPITest.cs:80]: consider validating the values of lower/upper bound as well.

File: src/OpenTelemetry/Metrics/MetricAggregators/HistogramMetricAggregator.cs [hunk 4]
```diff
@@ -62,11 +121,38 @@ internal HistogramMetricAggregator(string name, string description, string unit,
         public void Update<T>(T value)
             where T : struct
         {
-            // TODO: Implement Histogram!
+            // promote value to be a double
+
+            double val;
+            if (typeof(T) == typeof(long))
+            {
+                val = (long)(object)value;
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MetricAggregators/HistogramMetricAggregator.cs:129]: double or long?

File: test/OpenTelemetry.Tests/Metrics/MetricAPITest.cs [hunk 5]
```diff
@@ -14,20 +14,74 @@
...
 using System.Collections.Generic;
 using System.Diagnostics.Metrics;
-using System.Threading.Tasks;
 using Xunit;
 
-#nullable enable
-
 namespace OpenTelemetry.Metrics.Tests
 {
     public class MetricApiTest
```
Comment (cijothomas) [inline test/OpenTelemetry.Tests/Metrics/MetricAPITest.cs:24]: nit: I'd suggest to create a new class named HistogramAggregatorTest as this is purely testing the HistogramAggregator and nothing else.

**Review by cijothomas — APPROVED** [review]
LGTM.
left couple of small comments.

File: src/OpenTelemetry/Metrics/MetricAggregators/HistogramMetricAggregator.cs (hunk 2 above)
Comment (victlu) [reply src/OpenTelemetry/Metrics/MetricAggregators/HistogramMetricAggregator.cs:59]: Yes! We need 1 more than the size of the boundaries.

File: src/OpenTelemetry/Metrics/MetricAggregators/HistogramMetricAggregator.cs (hunk 4 above)
Comment (victlu) [reply src/OpenTelemetry/Metrics/MetricAggregators/HistogramMetricAggregator.cs:129]: We need to cast the value (which is a long) to a long. Then let compiler convert from long to double.

File: src/OpenTelemetry/Metrics/MetricAggregators/HistogramMetricAggregator.cs (hunk 2 above)
Comment (victlu) [reply src/OpenTelemetry/Metrics/MetricAggregators/HistogramMetricAggregator.cs:59]: But, I'll refactor to make it clearer!

File: src/OpenTelemetry/Metrics/MetricAggregators/HistogramMetricAggregator.cs (hunk 1 above)
Comment (victlu) [reply src/OpenTelemetry/Metrics/MetricAggregators/HistogramMetricAggregator.cs:137]: Sure

File: test/OpenTelemetry.Tests/Metrics/MetricAPITest.cs (hunk 3 above)
Comment (victlu) [reply test/OpenTelemetry.Tests/Metrics/MetricAPITest.cs:80]: It'll be a bit more complicated to do that, so, I like to skip this for now.

**Comment by cijothomas** [comment]
@victlu this can be closed in favor of the other PR towards main, right?

**Comment by victlu** [comment]
I'll close this PR as I have another PR for main branch.

**Pull request closed without merging**